### PR TITLE
Protocol Buffer text format

### LINKED ISSUE #5091: Text Based Protocol Buffer Support
The text based protocol buffer format does not highlight. It seems there was at least recognition of the elusive text based protocol buffer format back in https://github.com/github/linguist/issues/3816 when it was open.

I am unable to find any formal documentation on the format itself in the Protocol Buffer specification. However, there are methods within the Protocol Buffer API that handle encoding and decoding to and from the text based format.

You can use `protoc --encode` and `protoc --decode` on the command line to test.

Here is great example of the output format:
https://github.com/google/nvidia_libs_test/blob/master/cudnn_benchmarks.textproto

Some additional information:
https://gist.github.com/henridf/704c1c812f04a502c1c26f77a739090b

Currently I can mark the type as JSON5 and have somewhat acceptable highlighting based on the keys and values, however the format has a few deviations from what I can tell...
- keys do not need to be wrapped in an object, they can exist at the top level
- Comments are denoted by a `#`
- Commas are not used to denote an entry in an object, newlines are used in their place.
- Commas are used in array entries
- Objects do not use a colon for it's key name when defined. i.e `object { ... }` vs `object-key: { ... }`

## Preliminary Steps
Include a text file in your repository including a text encoded protocol buffer message with the extension `.proto` or mark the file extension in `.gitattributes` with ` linguist-language=Protocol-Buffer`.

Please confirm you have...
- [✔️] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
- [✔️] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
- [✔️] considered implementing an [override](https://github.com/github/linguist#overrides),
- [✔️] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).

## Problem Description
I am not necissarily advocating that the `Protocol-Buffer` type support this format under the entry for proto files, but rather advocating that there is a human readable, syntax highlight-able format that is emitted and would make it easier to read and diff were it supported. 

Protocol-Buffer-Text, textproto, or something of the like seems like the likely candidate for detection.

### URL of the affected repository:
https://github.com/ts-defold/tsd-template

### Expected language:
n/a

### Detected language:
n/a

### PR DESCRIPTION
**Pull request by hickford**
[Protocol Buffer text format](https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format) is a data format, typically parsed together with a Protocol Buffers schema. Thus Protocol Buffers text format is to Protocol Buffers as XML is to XSD.

TextMate Grammar by https://github.com/thejustinwalsh/textproto-grammar (MIT licence)

Fixes #5091.

## Checklist:

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      - `.pbtxt` 2000k hits https://github.com/search?q=extension%3Apbtxt&type=Code
      - `.textproto` 100k hits https://github.com/search?q=extension%3Atextproto&type=Code
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s): 
      - https://github.com/google/nvidia_libs_test/blob/master/cudnn_benchmarks.textproto  
      - https://github.com/tensorflow/tfx/blob/master/tfx/examples/chicago_taxi_pipeline/data/user_provided_schema/schema.pbtxt
    - Sample license(s): Apache-2.0 License, Apache-2.0 License
  
  - [x] I have included a syntax highlighting grammar: [example](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fgithub.com%2Fthejustinwalsh%2Ftextproto-grammar%2Fblob%2Fmain%2Fsyntaxes%2Ftextproto.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fgithub.com%2Fgoogle%2Fnvidia_libs_test%2Fblob%2Fmaster%2Fcudnn_benchmarks.textproto&code=)

### CODE REVIEW

**Comment by Alhadis** [comment]
First of all, please don't add `language_id`s manually. These need to be generated by running [`script/update-ids`](script/https://github.com/github/linguist/blob/1403100923bd45b33b12b369f998a1898c531140/script/update-ids).

Secondly, how does this format differ from the currently-supported [Protocol Buffer](https://github.com/github/linguist/blob/80f3531e8a1014a23f4606458e5a528053ed3cac/lib/linguist/languages.yml#L4664-L4675) language? Is it a different format, or should `.pbtxt` and `.textproto` be added to the existing entry's `extensions` list?

https://github.com/github/linguist/blob/80f3531e8a1014a23f4606458e5a528053ed3cac/lib/linguist/languages.yml#L4664-L4675

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -4673,6 +4673,17 @@ Protocol Buffer:
   codemirror_mode: protobuf
   codemirror_mime_type: text/x-protobuf
   language_id: 297
+Protocol Buffer text format:
```
Comment (Alhadis) [inline lib/linguist/languages.yml:4676]: **Nitpick:**

```suggestion
Protocol Buffer Text Format:
```

The capitalisation of `samples/Protocol Buffer text format` will need to be updated as well. Make sure your checkout's `core.ignorecsae` setting is disabled:

~~~sh
git config --local core.ignorecase false
~~~

(This has been a pitfall for many a contributor, including myself…)

**Review by Alhadis — CHANGES_REQUESTED** [review]

**Review by Alhadis — APPROVED** [review]
Looks good to me! 👍 Thanks for the superb contribution!

File: lib/linguist/languages.yml (hunk 1 above)
Comment (hickford) [reply lib/linguist/languages.yml]: Done.

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -4673,6 +4673,17 @@ Protocol Buffer:
...
   language_id: 297
+Protocol Buffer Text Format:
+  type: data
+  aliases:
+  - text proto
+  - protobuf text format
+  extensions:
+  - ".textproto"
+  - ".pbtxt"
+  tm_scope: none
```
Comment (Alhadis) [inline lib/linguist/languages.yml:4684]: Yikes, I just caught this.

```suggestion
  tm_scope: source.textproto
```

`tm_scope` needs to match the [`scopeName`](https://github.com/thejustinwalsh/textproto-grammar/blob/34f4ec92fac956da753c2818f765cb2575cba8de/syntaxes/textproto.tmLanguage.json#L305) property of the grammar to use for highlighting. It's basically a grammar's unique identifier.

**Review by Alhadis — CHANGES_REQUESTED** [review]

**Review by lildude — CHANGES_REQUESTED** [review]
Looks good, just one request: can you please add a sample for the `.pbtxt` extension.

**Comment by hickford** [comment]
Good idea, done.

**Review by lildude — APPROVED** [review]

**Comment by Stanzilla** [comment]
Could we also get .pbt added?

**Comment by lildude** [comment]
Sure, though that would now need to be a new PR as I've just merged this one.

**Pull request merged**